Windows Support for cpp_rpc

jmorrill opened the pull request:
Cherry picked PR from (PR #4548).

This adds Windows support to the C++ RPC server.
To enable building it make sure the CMake option USE_CXX_RPC is set to ON.

Notables:

- To workaround the lack for fork() sys call in Windows, the RPC socket is passed (WSADuplicateSocket magic) to a new child process.
- To workaround the lack of tar on Windows, wsl must be installed as tar is executed with "wsl tar -C..."
- Clang is used for compilation, so make sure[ LLVM binaries are installed ](http://releases.llvm.org/download.html)and in the system path.
- Updated CMakeLists.txt to use v3.9 on Windows to use INTERPROCEDURAL_OPTIMIZATION for better optimization of the libraries tvm.dll and topi.dll.  
- CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS is used on the tvm_runtime.dll so exports always work when using the graph runtime from C++ projects.

I have not verified I have not broken the Linux build of this in some way.  Certain liberties were also taken as I didn't expect to ever be making this PR, so apologies :)

To debug, I use these command args:

server --host=10.1.30.5 --port=9000 --port-end=9090  --tracker=10.1.30.5:9190 --key=1080ti

@soiferj @FrozenGene

FrozenGene commented:
Thanks @jmorrill to bring C++ RPC to Windows and change build system to CMake!

I maybe have no time to review it this week, but I have a glance in CMakeLists.txt. In our original C++ RPC Makefile, we will distinguish Android and Linux. Because we can't link pthread on Android, but we must link pthread on Linux. Seems you don't handle this?

jmorrill commented:
@FrozenGene Thanks for taking a peek.  I totally understand limited time.

You are right, I do not include a case for Android to skip linking with pthread.  My only excuse was can't really find a solid answer to this in CMake. "if(ANDROID)" would be nice if that's how it is done.  I'll be very happy to implement any suggestion though!
  
I made this "first class" option from CMake because a) Its a much better implementation on Windows than the python version, which took a bunch of hacks to make decent performance.  b) Was very easy to link to the tvm libs given its a sub CMake project.

FrozenGene commented:
> @FrozenGene Thanks for taking a peek. I totally understand limited time.
> 
> You are right, I do not include a case for Android to skip linking with pthread. My only excuse was can't really find a solid answer to this in CMake. "if(ANDROID)" would be nice if that's how it is done. I'll be very happy to implement any suggestion though!
> 
> I made this "first class" option from the main CMake because a) Its a much better implementation on Windows than the python version, which took a bunch of hacks to make decent performance. b) Was very easy to link to the tvm libs given its a sub CMake project.

I agree all about the advantages of CMake. For Android, I think the [CMAKE_SYSTEM_NAME](https://cmake.org/cmake/help/v3.4/variable/CMAKE_SYSTEM_NAME.html) maybe could help us. But it requires the users to specify the build os, I prefer we do it automatically, i.e. we could use the compiler option `-dumpmachine` to get we are in the android or linux.

For example, when I run:
` ~/ndk_tools/bin/clang -dumpmachine`, it will output: `aarch64-none-linux-android`
If I run arm linux compiler, it will output: `aarch64-linux-gnueabi`.

So I think we could this to distinguish android or linux. If we can not run this command successfully, we do the default things. 

Please make sure this [Android ndk tools](https://developer.android.com/ndk/downloads) and [linaro linux] (https://www.linaro.org/downloads/) could compile cpp/rpc successfully, because they are important platforms to use cpp_rpc. (Note. these two tools has windows os support too)

soiferj commented on apps/cpp_rpc/rpc_env.cc:
```diff
@@ -20,141 +20,139 @@
+#if defined(__linux__) || defined(__ANDROID__)
+  const std::string untar_cmd = "tar -C ";
+#elif defined(_WIN32)
+  const std::string untar_cmd = "wsl tar -C ";
```
I don't think it's ideal to have to force Windows users to enable WSL. Can we use Python [tarfile](https://docs.python.org/2/library/tarfile.html#examples) instead? Are there any other options?

jmorrill replied:
I considered some open-source C and C++ libraries, but shied away with license and time to implement reasons.

I think Python tarfile would be a good pick, especially if it comes packaged with python....as long as having python is not a burden.  I would like to test the speed though, mostly around python startup time and determine what is "acceptable" or not.

As a side note, it seems you can just send a precompiled dll (no tar), by changing the build_func:

`from tvm.contrib import cc`
....
`builder=autotvm.LocalBuilder(n_parallel=18, timeout=25, build_func=cc.create_shared)`

jmorrill replied:
I just did a few tests in powershell:
`Measure-Command {start-process "python tar-test.py" -Wait}`

Where tar-test.py is just: "import tarfile"

`Measure-Command {start-process "python wsl tar --help" -Wait}`

It doesnt look like python startup is much worse than "wsl tar".  Both around ~30ms.

How would you suggest this be implemented?  Generate a python script at runtime in the tmp folder?

jmorrill replied:
Looks like I do not need to create a python script at runtime.  It looks like something like this would work on windows:

python -m tarfile -e file.tar outdir/

tqchen commented:
@soiferj @FrozenGene please followup, and https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

tqchen commented:
@jmorrill @FrozenGene it would be great if we can followup :)

jmorrill commented:
> @jmorrill @FrozenGene it would be great if we can followup :)

Sorry about that.  I had changed around how the untar works based on @soiferj's suggestion.  But on @FrozenGene's concerns about android, it was beyond my expertise.  Basically, I could try to do it, but wouldn't even know where to begin to know if it were right :).

If everyone is too busy, I could withdraw the PR until a later time.

FrozenGene commented:
> > @jmorrill @FrozenGene it would be great if we can followup :)
> 
> Sorry about that. I had changed around how the untar works based on @soiferj's suggestion. But on @FrozenGene's concerns about android, it was beyond my expertise. Basically, I could try to do it, but wouldn't even know where to begin to know if it were right :).
> 
> If everyone is too busy, I could withdraw the PR until a later time.

For Android / Embed Linux, if we can not change to CMake easily, we could keep our current Makefile system.

jmorrill commented:
> For Android / Embed Linux, if we can not change to CMake easily, we could keep our current Makefile system.

I know visual studio has "nmake", which can work with Makefiles.  I can experiment with that and see if I can do similar to what is being done currently with the Makefile.

FrozenGene commented:
> > For Android / Embed Linux, if we can not change to CMake easily, we could keep our current Makefile system.
> 
> I know visual studio has "nmake", which can work with Makefiles. I can experiment with that and see if I can do similar to what is being done currently with the Makefile.

OK. This is one way. Another way is for Windows, we could use your pr's cmake. But for Android / Embed Linux, we could still use current Makefile system.

jmorrill commented:
Ok cool.  I'm do another commit in the next day or so as I think I can undo a lot of changes I made to the main CMake file and see what you guys think.

jmorrill commented:
Ok @FrozenGene I have made some changes to the CMakeLists.txt

Should I place "`tvm_option(USE_CPP_RPC "Build CPP RPC" OFF)`" with an "`if(WIN32)`" around it?  It will compile for linux, just it won't have the android options available in your Makefile.

Also I have:
```
if(WIN32)
  cmake_minimum_required(VERSION 3.9)
else()
  cmake_minimum_required(VERSION 3.2)
endif()

```

This gives a min cmake version of 3.9 on Windows.  It is not needed for this, but was hoping to sneak in interprocedural optimizations in Windows builds for extra perf. (ex `set_property(TARGET tvm PROPERTY INTERPROCEDURAL_OPTIMIZATION TRUE)` ) . 

I can remove that if its a problem and find a way to do it via compiler switches in a later PR.

FrozenGene commented:
> Ok @FrozenGene I have made some changes to the CMakeLists.txt
> 
> Should I place "`tvm_option(USE_CPP_RPC "Build CPP RPC" OFF)`" with an "`if(WIN32)`" around it? It will compile for linux, just it won't have the android options available in your Makefile.
> 
> Also I have:
> 
> ```
> if(WIN32)
>   cmake_minimum_required(VERSION 3.9)
> else()
>   cmake_minimum_required(VERSION 3.2)
> endif()
> ```
> 
> This gives a min cmake version of 3.9 on Windows. It is not needed for this, but was hoping to sneak in interprocedural optimizations in Windows builds for extra perf. (ex `set_property(TARGET tvm PROPERTY INTERPROCEDURAL_OPTIMIZATION TRUE)` ) .
> 
> I can remove that if its a problem and find a way to do it via compiler switches in a later PR.

I suggest when we build CPP_RPC on non-Windows platforms, we will raise one `fatal` and prompt users to use `app/Makefile` to build CPP RPC.

For CMake version, It is better to use the same version on Linux / Windows. We cold add it in compiler options in later pr.

jmorrill commented:
> For CMake version, It is better to use the same version on Linux / Windows. We cold add it in compiler options in later pr.

Removed all the extra CMakeLists.txt stuff I had in there and do a FATAL if USE_CPP_RPC is set on a non-Windows platform.  

Added instructions in the README.

Two other things I want to check with you guys is the small modification I made to `src/support/ring_buffer.h`.  The original would cause a crash on Windows.

Also the same with `src/runtime/rpc/rpc_socket_impl.cc` change, which shouldn't be an issue (put try/catch in a destructor).

FrozenGene commented:
> > For CMake version, It is better to use the same version on Linux / Windows. We cold add it in compiler options in later pr.
> 
> Removed all the extra CMakeLists.txt stuff I had in there and do a FATAL if USE_CPP_RPC is set on a non-Windows platform.
> 
> Added instructions in the README.
> 
> Two other things I want to check with you guys is the small modification I made to `src/support/ring_buffer.h`. The original would cause a crash on Windows.
> 
> Also the same with `src/runtime/rpc/rpc_socket_impl.cc` change, which shouldn't be an issue (put try/catch in a destructor).

i think it is fine

FrozenGene commented:
Thanks for the great work! @jmorrill This is merged now. Thanks for the review @soiferj !
